src/matchmaker.rs

```diff
@@ -0,0 +1,358 @@
+                    .query
+                    .push_str(format!("{}{}{}{}{}", boolean, field, op, value, boost).as_str())
+            }
+        }
+
+        self.matchmaker
+    }
+}
+
+impl<'a> Matchmaker {
+    pub fn new() -> Self {
+        Matchmaker {
+            min_count: 2,
+            max_count: 100,
+            string_properties: HashMap::new(),
+            numeric_properties: HashMap::new(),
+            query: "".to_owned(),
+        }
+    }
+
+    pub fn string_properties(&self) -> String {
+        let mut str = "{".to_owned();
+
+        let mut properties = self.string_properties.iter().map(|property| {
+            format!("\"{}\": \"{}\"", property.0, property.1)
+        }).collect::<Vec<String>>();
+
+        // The tests need to have a deterministic order
+        #[cfg(test)]
+        properties.sort();
+
+        str += &properties.join(",");
+
+        str += "}";
+        str
+    }
+
+    pub fn numeric_properties(&self) -> String {
+        let mut str = "{".to_owned();
+
+        let mut properties = self.numeric_properties.iter().map(|property| {
+            format!("\"{}\": {}", property.0, property.1)
+        }).collect::<Vec<String>>();
+
+        // The tests need to have a deterministic order
+        #[cfg(test)]
+        properties.sort();
+
+        str += &properties.join(",");
+
+        str += "}";
+        str
+    }
+
+    pub fn property_exists(&self, name: &str) -> bool {
+        self.string_properties.contains_key(name) || self.numeric_properties.contains_key(name)
+    }
+
+    pub fn min(&mut self, min: u32) -> &mut Self {
+        self.min_count = min;
+        self
+    }
+
+    pub fn max(&mut self, max: u32) -> &mut Self {
+        self.max_count = max;
+        self
+    }
+
+    pub fn add_string_property(&mut self, name: &str, value: &str) -> &mut Self {
+        assert!(!self.property_exists(name));
+
+        self.string_properties
+            .insert(name.to_owned(), value.to_owned());
+
+        self
+    }
+
+    pub fn add_numeric_property(&mut self, name: &str, value: f64) -> &mut Self {
+        assert!(!self.property_exists(name));
+
+        self.numeric_properties.insert(name.to_owned(), value);
+
+        self
+    }
+
+    pub fn add_query_item(&'a mut self, property: &'a str) -> QueryItemBuilder<'a> {
+        QueryItemBuilder {
+            matchmaker: self,
+            property,
+            boolean: Boolean::Optional,
+            query_type: None,
+            boost: 0,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn properties_formatting() {
+        let mut matchmaker = Matchmaker::new();
+        matchmaker.add_string_property("region", "Europe")
+            .add_string_property("host", "Windows")
+            .add_numeric_property("rank", 5.5)
+            .add_numeric_property("elo", 1000.0);
+
+        assert_eq!(matchmaker.string_properties(), "{\"host\": \"Windows\",\"region\": \"Europe\"}");
```

Comment:
I really really like how it works :+1: 
I suffered with hand-crafting nakama-queries quite a lot and this would help me so much!

Reply:
Glad you like it :)